### ISSUE TITLE
Simplify using UWP file pickers from desktop apps

### ISSUE THREAD

**Issue by Jaiganeshkumaran** [microsoft/WindowsAppSDK#1063]
To use UWP file pickers such as file open picker or file save picker in desktop apps, you need to cast the object reference to IInitializeWithWindow and invoke its Initialize method. Although one can use Win32 file pickers, those lack some features like the ability to use picker contract apps. To simplify usage, I would suggest adding an overload for the constructor of FileOpenPicker and FileSavePicker to take a window handle. Since HWND can't be represented with the Windows Runtime type system, it can take an int64_t instead.

**Comment by angelazhangmsft** [comment]
@Jaiganeshkumaran - as FYI in C#/.NET 5, we have just added support for several COM WinRT interop interfaces. Here are docs for more information: https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/winrt-com-interop-csharp

With these new APIs, you can add the following code in a C#/WinUI3 app which simplifies the code:
```
var hwnd = WinRT.Interop.WindowNative.GetWindowHandle(this);
var folderPicker = new Windows.Storage.Pickers.FolderPicker();
WinRT.Interop.InitializeWithWindow.Initialize(folderPicker, hwnd);
```

**Comment by Jaiganeshkumaran** [comment]
> @Jaiganeshkumaran - as FYI in C#/.NET 5, we have just added support for several COM WinRT interop interfaces. Here are docs for more information: https://docs.microsoft.com/en-us/windows/apps/desktop/modernize/winrt-com-interop-csharp
> 
> With these new APIs, you can add the following code in a C#/WinUI3 app which simplifies the code:
> 
> ```
> var hwnd = WinRT.Interop.WindowNative.GetWindowHandle(this);
> var folderPicker = new Windows.Storage.Pickers.FolderPicker();
> WinRT.Interop.InitializeWithWindow.Initialize(folderPicker, hwnd);
> ```

Right but that's still a lot of work compared to an UWP app where you just construct a picker and use it. Also these helpers must be rewritten for each supported programming language so I think it's better to have a constructor overload that does the initialization with window automatically.

**Comment by Scottj1s** [comment]
Agreed that this interop pattern is cumbersome and not discoverable. Ideally, the constructor overload would flow from a metadata attribute, [as was done for FastABI support](https://github.com/microsoft/cppwinrt/blob/464e59198c2046125d8ee5a8cc6ff1739097345f/cppwinrt/helpers.h#L201).  If a class was decorated with a CanInitializeWithWindowAttribute, a constructor could be generated to accept an HWND source object (i.e., implementing IWindowNative), from which the given class could then call IInitializeWithWindow.Initialize.  That would accrue to all language projections, rather just C#/WinRT.

**Comment by riverar** [comment]
Ideally, you'd just internalize all this goop in the `Window` class, give it a `HWND` property, and everything else just falls into place. It's unfortunate how complicated this has all become over what appears to be some misguided attempt to make `Window` "pure".

**Comment by Scottj1s** [comment]
As @Jaiganeshkumaran points out, it stems from HWND not being a WinRT type.  Such types require a backdoor (COM interop interface) to access.  But we haven't done a good job of supporting language projections there, and have effectively dropped WinRT/COM interop support on the floor.

**Comment by Scottj1s** [comment]
Out API design team is going to investigate making this interop pattern simpler (hopefully nonexistent in most cases).  Meantime, I've improved discoverability of the need for the pattern in C#/WinRT.  Failure to call IInitializeWithWindow now presents a diagnostic that points to an interop code sample to copy:
![image](https://user-images.githubusercontent.com/12356643/127583425-758f13fc-fb4e-4941-94c9-00f12c934a31.png)

**Comment by dotMorten** [comment]
@Scottj1s The improved error message is good see. I had some comments here: https://github.com/microsoft/microsoft-ui-xaml/issues/4167#issuecomment-889553620 that touched on some of this and used the pickers as an example.

**Comment by Scottj1s** [comment]
Thanks @dotMorten, and for linking the related issue.  Totally agree!

**Comment by DinahK-2SO** [comment]
Hi @jaigak, thanks for raising this topic. We’ve introduced the new `Microsoft.Windows.Storage.Pickers APIs` - feel free to take a look when you have a moment:
[FileOpenPicker(WindowId) Constructor](https://learn.microsoft.com/en-us/windows/windows-app-sdk/api/winrt/microsoft.windows.storage.pickers.fileopenpicker.-ctor?view=windows-app-sdk-1.8#microsoft-windows-storage-pickers-fileopenpicker-ctor(microsoft-ui-windowid))

With the new APIs a picker can be constructed like this:
```csharp
using Microsoft.Windows.Storage.Pickers;

var picker = new FileOpenPicker(this.AppWindow.Id);     // if no app window (like in a console app), use `new FileOpenPicker(default)`
var result = await picker.PickSingleFileAsync();
```
